README
= VoltRb: A gem client for VoltDB

VoltRB is a gem client for VoltDB that uses the JSON interface.

= Installing VoltRb

Get from RubyForge.

  $ gem install voltrb

= Example

  This is the Ruby version of VoltDB's hello world example.

  require 'voltrb'

  # Create a client and connect to the VoltDB instance running in the same machine.
  client = VoltRb::Client.new

  # Insert "Hello, World" for different languages.
  client.call_procedure("Insert", "Hello", "World", "English")
  client.call_procedure("Insert", "Bonjour", "Monde", "French")
  client.call_procedure("Insert", "Hola", "Mundo", "Spanish")
  client.call_procedure("Insert", "Hej", "Verden", "Danish")
  client.call_procedure("Insert", "Ciao", "Mondo", "Italian")

  # Say it in Spanish!
  response = client.call_procedure("Select", "Spanish")
  puts "#{response.results[0].first[:HELLO]}, #{response.results[0].first[:WORLD]}!"

= Usage

== Initialization

Initializing a client without passing any arguments will have it connect to the VoltDB instance running in localhost.

  client = VoltRb::Client.new

Or we can specify options as a hash.

  client = VoltRb::Client.new({:host => "voltdbhost", :port => 8888})

Right now, VoltRb only knows two:

* host - should point to the name or IP address of the VoltDB instance.
* port - in case the instance isn't listening on the default port (8080).

There are no options yet for authentication since the JSON interface doesn't implement any. See Notes for more.

== Stored Procedure Invocation

We call VoltDB stored procedures using Client#call_procedure. The first argument is the name of the procedure followed by its parameters if any.

  client.call_procedure(:Insert, "Hello", "World", "English")

Ruby types are automatically mapped to VoltDB types as below:

Ruby::    VoltDB 
Fixnum, Bignum::  Tiny, Small, Integer, Long
Float::   Float, Double
BigDecimal:: Decimal
Time, Date, DateTime::  Timestamp
String::  String

Nils become Nulls. 

== Stored Procedure Response

VoltDB stored procedures can return multiple rowsets (VoltTable objects). This array of rowsets is returned by ProcedureResponse#results. In the hello world example, the Select procedure returns only one so:

  response = client.call_procedure("Select", "Spanish")
  response.results[0]
  => #<VoltRb::VoltTable:0x17952c4 @raw={"status"=>-128, "schema"=>[{"name"=>"HELLO", "type"=>9}, 
  {"name"=>"WORLD", "type"=>9}], "data"=>[["Hola", "Mundo"]]}>

Each VoltTable / rowset implements Enumerable so we can access the rows using the usual methods #each, #all, #first, #last.

  response.results[0].each { |row| puts row }

Each row is a hash with keys that are symbolized column names.

  response.results[0].each { |row| puts "#{row[:HELLO]}, #{row[:WORLD]}" }
  Hola, Mundo
  => [["Hola", "Mundo"]]

You can get the list of columns by looking at the array returned by VoltTable#schema.

  response.results[0].schema
  => [#<struct VoltRb::VoltColumn name=:HELLO, type=9>, #<struct VoltRb::VoltColumn name=:WORLD, type=9>]

== Catching Errors

Standard Ruby errors will be raised in case of network connection problems. VoltDB-specific errors like invalid procedure arguments and constraint violations will raise VoltRb::VoltError. We can find out more by inspecting #status_string or #app_status_string.

  begin
    client.call_procedure("Insert", "Hello", "World", "English")
    # This second call will cause a unique constraint violation.
    client.call_procedure("Insert", "Hello", "World", "English")
  rescue VoltRb::VoltError => bang
    puts "Error: #{bang.status_string}"
  end    

  Error: 
  ========================================================
  VOLTDB ERROR: CONSTRAINT VIOLATION
    Attempted violation of constraint
  Constraint Type UNIQUE, Table CatalogId HELLOWORLD
   header size: 35
   status code: -128 column count: 3
   cols (HELLO:STRING), (WORLD:STRING), (DIALECT:STRING), 
   rows -
    Hello, World, English, 
  
      at Insert.run(Insert.java:19)
  ========================================================
 
= Notes

== User Authentication

As of VoltDB v1.1, the JSON interface does not implement any form of authentication. 

The following text was taken from VoltDB's JSON documentation:

<em>Note that there is currently no way to authenticate the client interface using JSON. In other words, the JSON interface operates as an "anonymous" client. If security is enabled for the database, use of the JSON interface is essentially disabled because any attempt to execute a stored procedure will result in a protection violation.</em>

This forum thread says they expect to have the feature in v1.2

http://community.voltdb.com/node/241

== Unicode Strings

As of v1.1 VoltDB's JSON interface does not properly return UTF-8 strings so non-Latin characters don't display correctly.

http://community.voltdb.com/node/263

= License

This gem has been released under the MIT License[link:files/MIT-LICENSE.html].

= Warranty

This software is provided "as is" and without any express or
implied warranties, including, without limitation, the implied
warranties of merchantibility and fitness for a particular
purpose.

== Contributing

The source code is hosted on http://github.com/beljun/voltrb.

== Credits

Author: Junjun Olympia (@beljun) / http://rubyredtomatoes.com




CHANGELOG
= Change log

== 0.0.1
* Initial release.

== 0.0.2
* Updated the README about the Unicode string issue. The trunk version now has this fixed.
* More robust handling of different date & time types.

== 0.1.0
* Client now accepts username/password to authorize connection with VoltDB servers (>= v1.2)



lib/voltrb/client.rb
require 'json'
require 'rest_client'
require 'date'
require 'time'
require 'bigdecimal'
require 'digest/sha1'

module VoltRb
  # This is the main class we use to interact with the VoltDB instance.
  #
  # == Initialization
  # Create a client and connect to the instance in localhost:
  #   client = VoltRb::Client.new
  #
  # Or pass it a hash of options like:
  #   client = VoltRb::Client.new({:host => "voltdb_server", :port => 8888})
  #
  # See the new method for more details.
  #
  # == Calling stored procedures
  # Call a stored procedure:
  #   client.call_procedure(:MyStoredProc, "This is the first argument", 2, Date.today)
  #
  # The call above invokes the VoltDB stored procedure named "MyStoredProc". 
  # The procedure expects 3 parameters (a String, a small integer, and a Date/Timestamp) so we pass these in order.
  # See call_procedure method for more.
  #
  # == Getting the stored procedure response
  # call_procedure returns a ProcedureResponse object which will contain the rows returned if any:
  #   response = client.call_procedure(:AnotherProc, 1)
  #   puts response.results[0].first[:TITLE]
  #
  # In the above example, we invoke the procedure "AnotherProc" and expect one rowset and 
  # one row with a column named "TITLE".
  # See ProcedureResponse for more.
  #
  # == Handling errors
  # In case of typical errors like network connection problems, standard Ruby exceptions will be raised. 
  # In case of VoltDB-specific errors like invalid stored procedure arguments or constraint violations, 
  # a VoltDB::VoltError exception object will be raised.
  #   begin
  #     client.call_procedure(:MyStoredProc, 2)
  #   rescue VoltDB::VoltError => bang
  #     puts bang.status_string
  #     raise
  #   end
  #
  # The above call will raise an error because it expects 3 parameters but was only given one. 
  # We can find the details of the error by inspecting status_string or app_status_string.
  # See VoltError for more.
  class Client
    attr_reader :host, :port, :username
    attr_accessor :api_root

    # Calling new without any arguments will create a client and have it connect to the VoltDB instance running 
    # in localhost:8080.
    # Pass an options hash to modify these defaults. 
    #   client = VoltRb::Client.new({:host => "voltdb_server", :port => 8888})
    #
    # The following options are recognized:
    # * host - the machine name or IP address
    # * port - other than the default
    # * username 
    # * password
    #
    def initialize(options = {})
      @host = options[:host] || "localhost"
      @port = options[:port] || 8080
      @username= options[:username]
      @hashed_password = Digest::SHA1.hexdigest(options[:password]) if options[:password]
      @api_root = "http://#{@host}:#{@port}/api/1.0/"
    end

    # Invoke a VoltDB stored procedure by calling this method with the first argument being the stored procedure name. 
    # Follow that with any number of parameters needed by the stored procedure itself.
    #   client.call_procedure(:MyStoredProc, "This is the first argument for :MyStoredProc", 2, Date.today)
    #
    # Ruby types map to the stored procedure data types as below:
    # Ruby::    VoltDB 
    # Fixnum, Bignum::  Tiny, Small, Integer, Long
    # Float::   Float, Double
    # BigDecimal:: Decimal
    # Time, Date, DateTime::  Timestamp
    # String::  String
    # 
    # Nils become Nulls. 
    # If a stored procedure returns one or more rowsets that we're interested in, see the ProcedureResponse object
    # that's returned by this method.
    # In case of errors, handle the usual Ruby errors and the VoltDB-specfic VoltError (this inherits from StandardError).
    def call_procedure(procedure_name, *args)
      params = args.inject([]) { |o,e| o << prep_param(e); o }
      payload = {:Procedure => procedure_name.to_s, :Parameters => params.to_json}
      if @username
        payload[:User] = @username
        payload[:Hashedpassword] = @hashed_password
      end 
      response = RestClient.post(@api_root, payload, :content_type => "text/plain; charset=UTF-8", :accept => :json)
      proc_resp = ProcedureResponse.new(response)
      raise(VoltError.new(proc_resp.status, proc_resp.status_string, proc_resp.app_status, proc_resp.app_status_string), "A VoltDB procedure error occurred. See the exception object for details.") if proc_resp.raw["status"] != 1 or proc_resp.raw["appstatus"] != -128
      proc_resp
    end

    # Alias for call_procedure.  Term used in other client libraries.
    alias :invoke :call_procedure

  private
    def prep_param(val)
      case val
        when BigDecimal then val.to_s('F')
        when Time then val.to_i * 1000
        # Note: case DateTime MUST come before Date because Ruby seems to think DateTime is a Date. But Date is not a DateTime. Ruby bug?
        when DateTime then Time.parse(val.strftime("%a %b %d %H:%M:%S %z %Y")).to_i * 1000
        when Date then Time.parse(val.strftime("%b %d %Y")).to_i * 1000
        else val
      end
    end
  end
end

